RSSTimeFormatter/RealDateTimeFormatter.cs

```diff
@@ -6,257 +6,287 @@
-			}
-			DateTime epoch = GetEpoch();
-			DateTime target = epoch.AddSeconds(time);
-			TimeSpan span = target - epoch;
-			return string.Format("{0}{1}{2}{3}{4}"
-				, isNegativeTime ? "- " : (explicitPositive ? "+ " : "")
-				, (valuesOfInterest >= 3 && span.Days != 0) ? string.Format("{0}d, ", span.Days) : ""
-				, (valuesOfInterest >= 2 && span.Hours != 0) ? string.Format("{0}h, ", span.Hours) : ""
-				, (valuesOfInterest >= 1 && span.Minutes != 0) ? string.Format("{0}m, ", span.Minutes) : ""
-				, valuesOfInterest >= 0 ? string.Format("{0}s", span.Seconds) : ""
-			);
-		}
+        #region IDateTimeFormatter implementation
+        public string PrintTimeLong(double time)
+        {
+            // short-circuit if invalid time passed
+            if (IsInvalidTime(time))
+                return InvalidTimeStr(time);
+            TimeSpan span = TimeSpan.FromSeconds(time);
+            return spanAsLongFormDateString(span);
+        }
+        public string PrintTimeStamp(double time, bool days = false, bool years = false)
+        {
+            // short-circuit if invalid time passed
+            if (IsInvalidTime(time))
+                return InvalidTimeStr(time);
+            TimeSpan span = TimeSpan.FromSeconds(time);
+            return string.Format("{0}{1}"
+                , days ? string.Format("Day {0} - ", span.Days) : ""
+                , spanAs24HWithSeconds(span)
+            );
+        }
+        public string PrintTimeStampCompact(double time, bool days = false, bool years = false)
+        {
+            // short-circuit if invalid time passed
+            if (IsInvalidTime(time))
+                return InvalidTimeStr(time);
+            TimeSpan span = TimeSpan.FromSeconds(time);
+            int totalYears = (int)(span.TotalDays / 365);
+            return string.Format("{0}{1}{2}"
+                , years ? string.Format("{0}y, ", totalYears) : ""
+                , days ? string.Format("{0}d, ", span.Days) : ""
+                , spanAs24HWithSeconds(span)
+            );
+        }
+        public string PrintTime(double time, int valuesOfInterest, bool explicitPositive)
+        {
+            // This is a downright strange and confusing method but as I understand it
+            // what it is saying is give it the time in the following format:
+            // 1y, 1d, 1h, 1m, 1s
+            // But the param 'valuesOfInterest' is used to control how 'deep' it goes
+            // IOW a valueofInterest of say 3 would only give you hours, minutes, and seconds
+            // Why it isn't more straightforward is beyond me
+            // short-circuit if invalid time passed
+            if (IsInvalidTime(time))
+                return InvalidTimeStr(time);
+            bool isTimeNegative = time < 0;
+            time = Math.Abs(time);
+            TimeSpan span = TimeSpan.FromSeconds(time);
+            int years = (int)(span.TotalDays / 365);
+            return string.Format("{0}{1}{2}{3}{4}{5}"

```

Comment:
If `valuesOfInterest` is 1 and you have a time of less than a day, this will not print anything. I don't think this is intended behaviour?

Reply:
Well, I don't know. Depends on how the original KSP interface is designed I guess.
If `valuesOfInterest` is supposed to set the significance of the answer, this would be correct. If you're only interested in the amount of years, a length of less than a year should be either '0' or maybe '<1' .
If `valuesOfInterest` is supposed to set the amount of fields returned it is indeed wrong. If the latter is the case, the fix would be easy, but it depends on the intended design.

Reply:
There is an issue that you don't get anything if the number of years (/months/etc) is less than one. `PrintTime(1, 1, false)` yields an empty string for example, not "0".

Reply:
The original KSP code (or at least a version made from decompile) is here for reference: https://github.com/KSP-RO/RSSTimeFormatter/blob/master/RSSTimeFormatter/DefaultDateTimeFormatter.cs#L82-L109

Reply:
I've changed `PrintTime()` to be like the original KSP code. I.e. it will give `valuesOfInterest` amount of values from the first non-zero value.
e.g.:
```
PrintTime("0y 1d 2h 3m 4s", 0, false) = "1d"
PrintTime("0y 0d 2h 3m 4s", 0, false) = "2h"
PrintTime("0y 1d 2h 3m 4s", 1, false) = "1d, 2h"
PrintTime("1y 0d 0h 3m 4s", 1, false) = "1y, 0d"
PrintTime("1y 0d 0h 3m 4s", 2, false) = "1y, 0d, 0h"
```